### ISSUE TITLE
HIGH: Deep dive on background sync by adding "dev tools" to native app to diagnose realtime Pusher/UrbanAirship problems on staging

### ISSUE THREAD

**Issue by quinthar** [Expensify/App#33040]
**Strategy:**
The key to WhatsApp quality (the gold standard to which we strive) is total reliability across all networking conditions, across all devices.

**Problem:**
I am active on both desktop and mobile.  However, when I open my mobile app after a long day of work, it shows a lot of read messages as unread.  I have reliably signal and wifi throughout, and I would expect it to be up to date when I open, but it seems out of date until it completes a "sync".  This has proven difficult to reproduce in testing because there are many different ways for this same basic problem to occur, spread across many third party systems (including Urban Airship and Pusher) that are difficult to test in dev.  Furthermore, we are limited by the difficulty of accessing device logs to reconstruct a clear timeline of actions across client and server.

**Solution:** 
Let's upgrade our test environment to dig into this deeper than we have been able to before.  Specifically:
1. Add a "Dev tools" section to the app that is enabled when the staging server is selected
2. One of the features simply shows all historical device console logs, including those that happen while the app is backgrounded or killed (ie, when it is woken up for push messages)
3. Write an end-to-end test script that walks through the process of reading a message on the server, and then verifying it shows up as read on the client, for various combinations of client being in the foreground, background, and killed
4. Map out what API calls should be happening for each of these flows -- and if in the process we realize the design is flawed, create issues to fix it.
5. Use the new dev tools to verify on the device when running the staging build on the staging server (ie, not in dev) that the API calls are happening in the correct order -- and wherever they are not, create issues to fix them.

**Comment by muttmuure** [comment]
@TMisiukiewicz this is the issue I messaged you about!

**Comment by quinthar** [comment]
Upgrading to HIGH as it's about making existing functionality work reliably, not just about new functionality.

**Comment by muttmuure** [comment]
@TMisiukiewicz now we've passed the holiday break, are you able to take a look at this in more detail? 

@quinthar posted a proposal here: https://expensify.slack.com/archives/C01GTK53T8Q/p1703982689203629 

@tomekzaw do you think Callstack could get a team together to work on this?

**Comment by TMisiukiewicz** [comment]
@muttmuure I'm currently quite busy with audit-related stuff and this will probably continue next week as well, but I am happy to participate in this once I finish all my current work 👍 would be great if someone else could take a deeper look on it somewhere next week

**Comment by TMisiukiewicz** [comment]
Hi @muttmuure what is the status here? I am currently available to work on this issue, however i see it's on hold

**Comment by muttmuure** [comment]
Would you be interested in taking on that linked issue? It doesn't look like Fedi has started it, and we're waiting on someone to get started

**Comment by muttmuure** [comment]
I think @quinthar is piecing this one out into microsteps. There's a good discussion to help you catch up here: https://expensify.slack.com/archives/C01GTK53T8Q/p1704796266475999?thread_ts=1703982689.203629&cid=C01GTK53T8Q

**Comment by TMisiukiewicz** [comment]
@muttmuure thanks! I can see Fedi is assigned as a C+, do you want me to review the proposals or start working on the proper solution?

**Comment by muttmuure** [comment]
I think starting on a solution would be welcomed

**Comment by muttmuure** [comment]
Do you mind if I circle back in a couple of days after I've run through the
testing steps I need to take, and executed some JS snippets?

**Comment by TMisiukiewicz** [comment]
@muttmuure no problem 👍

**Comment by TMisiukiewicz** [comment]
@muttmuure since the code is minified, I think at this moment it is not possible to filter with a JS snippet. As @tgolen said in the linked comment, we can consider exposing `Onyx` as a global variable that can be accessed. Then we could filter with a snippet, however I think it would be better to have a filtering option separately, as it would be easily available for everyone, regardless of the JS knowledge 🤔

**Comment by muttmuure** [comment]
OK, I think we're good for now. I will finish this up by writing up how reliable updates should work for the current workflow ([here](https://expensify.slack.com/archives/C0671JDRKQW/p1709231033088219?thread_ts=1708058905.414929&cid=C0671JDRKQW)) and then hand off for regular regression testing. 

We've been discussing next steps here: https://expensify.slack.com/archives/C0671JDRKQW/p1708058905414929

If anyone can answer my question here: https://expensify.slack.com/archives/C01GTK53T8Q/p1709220371691039?thread_ts=1709219747.260089&cid=C01GTK53T8Q I'd be grateful too as sometimes what is being logged is only half the story versus what the actual Onyx data contains
